boards/components/src/panic_button.rs

```diff
@@ -0,0 +1,79 @@
+//! Component to cause a button press to trigger a kernel panic.
+//!
+//! This can be useful especially when developping or debugging console
+//! capsules.
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! components::panic_button::PanicButtonComponent::new(
+//!     &sam4l::gpio::PC[24],
+//!     kernel::hil::gpio::ButtonMode::LowWhenPressed,
+//!     kernel::hil::gpio::FloatingState::PullUp
+//! ).finalize(());
+//! ```
+
+use kernel::component::Component;
+use kernel::hil::gpio;
+use kernel::static_init;
+
+pub struct PanicButtonComponent<'a> {
+    pin: &'a dyn gpio::InterruptPin,
+    mode: gpio::ButtonMode,
+    floating_state: gpio::FloatingState,
+}
+
+impl<'a> PanicButtonComponent<'a> {
+    pub fn new(
+        pin: &'a dyn gpio::InterruptPin,
+        mode: gpio::ButtonMode,
+        floating_state: gpio::FloatingState,
+    ) -> Self {
+        PanicButtonComponent {
+            pin,
+            mode,
+            floating_state,
+        }
+    }
+}
+
+impl Component for PanicButtonComponent<'static> {
+    type StaticInput = ();
+    type Output = ();
+
+    unsafe fn finalize(self, _: Self::StaticInput) -> Self::Output {
+        let panic_button = static_init!(
+            PanicButton,
+            PanicButton::new(self.pin, self.mode, self.floating_state)
+        );
+        self.pin.set_client(panic_button);
+    }
+}
+
+struct PanicButton<'a> {
```

Comment:
I think this should go in a capsule rather than directly in this component. To my knowledge, components have so far only been used for initialization logic, not for any actual code.

Reply:
Thanks! This replies to my question in the first message.

Reply:
However, my take on it is that splitting this file is unnecessary overhead. This requires making the `PanicButton` public, and documenting it, when it should be an internal detail and should only be used externally through the component. But I guess the same arguments would apply to `DebugProcessRestart` - which is its own capsule.

In general, I think it would even make sense to define components directly in each capsule as the only public API - except that it would require `unsafe` because `static_init` is unsafe, and capsules are currently `forbid(unsafe)` - which prevents such a pattern. So we're left with some overhead in splitting code between capsules and components.